### PR TITLE
NODE-73: Get account balance after application of transactions in UTX pool

### PR DESCRIPTION
**Pull request by vsuharnikov**
You can check a portfolio (including the balance) through a debug API-method:
```
/debug/portfolios/{address}?considerUnspent=true|false
```
, but don't forget to provide an API key ;)

* considerUnspent=*true* - including transactions from UTX pool;
* considerUnspent=*false* - without it.

### CODE REVIEW

File: src/it/scala/com/wavesplatform/it/api/NodeApi.scala [hunk 1]
```diff
@@ -56,6 +59,12 @@ trait NodeApi {
   def get(path: String, f: RequestBuilder => RequestBuilder = identity): Future[Response] =
     retrying(f(_get(s"http://$restAddress:$nodeRestPort$path")).build())
 
+  def getDebug(path: String, f: RequestBuilder => RequestBuilder = identity): Future[Response] = retrying {
```
Comment (ismagin) [inline src/it/scala/com/wavesplatform/it/api/NodeApi.scala]: `getWithApiKey` sounds more reasonable

File: src/main/scala/com/wavesplatform/UtxPool.scala [hunk 2]
```diff
@@ -2,22 +2,25 @@ package com.wavesplatform
...
-import scorex.transaction.{FeeCalculator, History, Transaction, ValidationError}
+import scorex.transaction._
 import scorex.utils.{ScorexLogging, Time}
 
 import scala.collection.JavaConverters._
 import scala.concurrent.duration._
 import scala.util.{Left, Right}
 
-
+/**
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/UtxPool.scala]: Please remove unnnessary comments

File: src/main/scala/com/wavesplatform/UtxPool.scala [hunk 3]
```diff
@@ -26,21 +29,55 @@ class UtxPool(
...
+
   private val transactions = new ConcurrentHashMap[ByteStr, Transaction]
+
   private lazy val knownTransactions = CacheBuilder.newBuilder()
     .maximumSize(utxSettings.maxSize * 2)
     .build[ByteStr, Either[ValidationError, Transaction]]()
 
-  private def removeExpired(currentTs: Long): Unit =
-    transactions.entrySet().removeIf(tx => (currentTs - tx.getValue.timestamp).millis > utxSettings.maxTransactionAge)
+  /**
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/UtxPool.scala]: The best comment is readable code :)

Comment (vsuharnikov) [inline src/main/scala/com/wavesplatform/UtxPool.scala]:
```scala
private val txPessimisticPortfolios = new ConcurrentHashMap[ByteStr, Portfolios]
```
I know, but it is hard to name this variable, i did my best. The comments here is for clearing things. 

If someone has an idea about naming, you're welcome :)

File: src/main/scala/com/wavesplatform/state2/Portfolio.scala [hunk 4]
```diff
@@ -7,6 +7,20 @@ import cats.Monoid
...
   lazy val spendableBalance: Long = balance - leaseInfo.leaseOut
+
+  /**
+    * The Portfolio with only the withdraw part
+    */
+  def pessimistic: Portfolio = Portfolio(
+    balance = Math.min(balance, 0),
+    leaseInfo = LeaseInfo(
+      leaseIn = 0,
+      leaseOut = Math.min(leaseInfo.leaseOut, 0)
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/state2/Portfolio.scala]: Shouldn't this be `Math.max(leaseInfo.leaseOut, 0)`
If you lease something, it's positive(should be included), if you get your lease back, it's negative(should not be included)

File: src/main/scala/com/wavesplatform/UtxPool.scala [hunk 5]
```diff
@@ -26,21 +26,55 @@ class UtxPool(
...
       Left(GenericError("Transaction pool size limit is reached"))
     } else knownTransactions.get(tx.id, () => {
       val validationResult = for {
         _ <- feeCalculator.enoughFee(tx)
-        _ <- TransactionDiffer.apply(fs, history.lastBlock.map(_.timestamp), time.correctedTime(), stateReader.height)(stateReader, tx)
+        diff <- TransactionDiffer(fs, history.lastBlock.map(_.timestamp), time.correctedTime(), stateReader.height)(stateReader, tx)
+        _ = {
+          val nonEmptyPessimisticPortfolios = diff.portfolios
+            .mapValues(_.pessimistic)
+            .filterNot {
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/UtxPool.scala]: Combining with empty does nothing, please simplify

File: src/main/scala/com/wavesplatform/UtxPool.scala [hunk 6]
```diff
@@ -55,6 +89,23 @@ class UtxPool(
     }
   }
 
+  def portfolio(addr: Address): Portfolio = {
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/UtxPool.scala]: Here you reduce ALL transaction diffs on every request.
Condider `Map[Address,List[TransactionId]]` to be able to fold only nessessary txs.

File: src/main/scala/com/wavesplatform/state2/Portfolio.scala [hunk 7]
```diff
@@ -7,6 +7,20 @@ import cats.Monoid
...
+    ),
+    assets = assets.filter { case (_, v) => v < 0 }
+  )
+
+  def isEmpty: Boolean = this == Monoid.empty[Portfolio]
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/state2/Portfolio.scala]: Not needed

File: src/main/scala/scorex/waves/http/DebugApiRoute.scala [hunk 8]
```diff
@@ -74,6 +76,40 @@ case class DebugApiRoute(settings: RestAPISettings,
...
+      value = "Taking into account pessimistic transactions from UTX pool",
+      required = false,
+      dataType = "boolean",
+      paramType = "query",
+      defaultValue = "true"
+    )
+  ))
+  @ApiResponses(Array(new ApiResponse(code = 200, message = "Json portfolio")))
+  def portfolios: Route = path("portfolios" / Segment) { (rawAddress) =>
+    (get & parameter('considerUnspent.as[Boolean])) { (considerUnspent) =>
```
Comment (ismagin) [inline src/main/scala/scorex/waves/http/DebugApiRoute.scala]: Not sure if anyone needs it, but shouldn't we also include `effectiveBalance` and the needed `spendableBalance`  to the response?

File: src/test/scala/com/wavesplatform/UtxPoolSpecification.scala [hunk 9]
```diff
@@ -49,6 +49,12 @@ class UtxPoolSpecification extends FreeSpec
   } yield TransferTransaction.create(None, sender, recipient, amount, time.getTimestamp(), None, fee, Array.empty[Byte]).right.get)
     .label("transferTransaction")
 
+  private def transferWithRecipient(sender: PrivateKeyAccount, recipient: PrivateKeyAccount, maxAmount: Long, time: Time) = (for {
```
Comment (ismagin) [inline src/test/scala/com/wavesplatform/UtxPoolSpecification.scala]: for `recipient`, `PublicKeyAccount` is enough.

File: src/it/scala/com/wavesplatform/it/transactions/debug/DebugPortfoliosSpecification.scala [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    val firstAddr = Address.fromString(firstAddress).right.get
+
+    val f = for {
+      _ <- assertBalances(firstAddress, 100.waves, 100.waves)
+      _ <- assertBalances(secondAddress, 100.waves, 100.waves)
+
+      portfolioBefore <- sender.debugPortfoliosFor(firstAddr, considerUnspent = true)
+      utxSizeBefore <- sender.utxSize
+
+      _ <- sender.payment(firstAddress, secondAddress, 5.waves, fee = 5.waves)
```
Comment (ismagin) [inline src/it/scala/com/wavesplatform/it/transactions/debug/DebugPortfoliosSpecification.scala]: Consider  `LeaseTransaction` here. This way we'll check `leaseOut component as well.

File: src/main/scala/com/wavesplatform/UtxPool.scala (hunk 6 above)
Comment (vsuharnikov) [inline src/main/scala/com/wavesplatform/UtxPool.scala]: With this approach we should go over all addresses in `UtxPool.removeExpired`:
```scala
      expiredTxs.foreach { tx =>
        transactions.remove(tx.id)
        txPessimisticPortfolios.remove(tx.id)
      }
```
but yes, I agree, that `removeExpired` could be executed rarely comparing with `portfolio` .

File: src/main/scala/com/wavesplatform/UtxPool.scala (hunk 5 above)
Comment (vsuharnikov) [inline src/main/scala/com/wavesplatform/UtxPool.scala]: > Combining with empty does nothing

That's why we filter out all empty portfolios :)
```scala
            .filterNot {
              case (_, portfolio) => portfolio.isEmpty
            }
```

File: src/main/scala/com/wavesplatform/state2/Portfolio.scala (hunk 7 above)
Comment (vsuharnikov) [inline src/main/scala/com/wavesplatform/state2/Portfolio.scala]: Please, see above

File: src/main/scala/scorex/waves/http/DebugApiRoute.scala (hunk 8 above)
Comment (vsuharnikov) [inline src/main/scala/scorex/waves/http/DebugApiRoute.scala]: I think, QAs should make this calculation by themselves, it is not too hard. If we really need this, i'd rather create a special API case class and make a [de]serializers for it.

File: src/it/scala/com/wavesplatform/it/transactions/debug/DebugPortfoliosSpecification.scala (hunk 10 above)
Comment (vsuharnikov) [inline src/it/scala/com/wavesplatform/it/transactions/debug/DebugPortfoliosSpecification.scala]: I added two tests for `Portfolio.pessimistic` in `PortfolioTest` to check `leaseOut` and other fields. I think, this is enough.

**Comment by vsuharnikov** [comment]
Thanks for helpful comments! Answered on some questions. Also please take a look onto `UtxPool`, a lot of changes are here.

**Pull request merged**